### ISSUE TITLE
"SWS_Wait until end of loop" not totally reliable

### ISSUE THREAD

**Issue by lowlou** [reaper-oss/sws#1463]
Hello 

I use SWS Wait until end of loop to automatically stop recordings at the end of set loop. 
But quite often, the action doesn't seem to work, instead what happens is that the recording stops, very quickly after starting (1/4 , 1/2 second, very fast).

here is the sequence of action in place (this is happening in the middle of a bigger cycle action): 

step 1 - Transport : Start/stop recording at edit cursor
step 2 - Wait until end of loop (if playing)
step 3 - Transport : Start/stop recording at edit cursor

It will sometimes work, it waits end of loop to stop.  Other times not, recording is canceled as soon as it starts.

I tried a script by Nofish before step 2, to implement a delay, thinking that maybe Reaper takes a little delay to start recording/"playing", but no luck. (Script is : "nofish_run action - wait 1 sec - run next action.lua" using "no-op" as the two actions embeded)

I don't know if something could be done to improve this ?

**Comment by nofishonfriday** [comment]
Haven't tested this yet, but maybe it helps if you put one of the native actions "Wait ... seconds before next action" after step 1?

**Comment by lowlou** [comment]
I tried things like this also, but I will try again :) !

**Comment by lowlou** [comment]
No, I just tested... Still happens.

The cycle action that houses the 3 step sequence is used via a script (the script is a mic of simple reaper commands, embeded cycle actions, and very simple lua logic : IF & ELSE kind of things ). I tried many things, like... Cut the cycle action in two (split point being between step one and two) and launch the 1seconds delay from Reaper or your delay-script, from the script itself...

LONG STORY SHORT, I tried many many combination (my action is _slick_ otherwise so I really wanted it to be perfect), but there is still a 1/4 chance ratio that the script will not wait until end of loop before stopping. Maybe I should copy paste your delayer script inside my own script ? Would it make a difference, rather than having a command_id ?

**Comment by nofishonfriday** [comment]
> (Script is : "nofish_run action - wait 1 sec - run next action.lua" using "no-op" as the two actions embeded)
> Maybe I should copy paste your delayer script inside my own script ? Would it make a difference, rather than having a command_id ? I'll test this.

if you have "no-op" as the two actions embedded it wouldn't have any effect I imagine, what you maybe could try as the embedded actions is:
1st action: Transport : Start/stop recording at edit cursor
2nd action: Wait until end of loop (if playing)

replacing steps 1+2 of your sequence above.

**Comment by lowlou** [comment]
I think I tried this also, but not 100% sure so I'll check again, thanks !!

**Comment by lowlou** [comment]
Your last suggestion doesn't work either :/ : I don't see any recording starting (no UI refresh indicating any recording during the action) and I end up with no recording, but the action seems to go thru all its steps.

**Comment by nofishonfriday** [comment]
I can reproduce this if the edit cursor is at the beginning of the project (couldn't reproduce if it is at other positions so far).

I've made a custom action consisting of your 3 actions from the first post. As can be seen in the screencapture, when running it and edit cursor  is at project start, the record button just flashes briefly, when I move the edit cursor forward a bit and try again (

When I move the edit cursor forward a bit and run the custom action again (0:18) it works as expected.
![#1463](https://user-images.githubusercontent.com/7658194/100380074-65302580-3016-11eb-8049-3b522b1e9a8a.gif)

Not sure though if this is an issue on SWS or Reaper side.

**Comment by lowlou** [comment]
Ok thx for your investigation. Sorry but on your gif I don't really see the cursor going back quickly. Seems like you do it with the transport bar ? Nevermind. Here, the problem is at any cursor position. Before the trio of action we're discusing, inside the cycle action, the cursor is moved by actions such as "go to start of time selection", "move cursor back one beat(no seak)". Probably not related but well... I'll have to live with it I guess ^^.

**Comment by nofishonfriday** [comment]
> Sorry but on your gif I don't really see the cursor going back quickly.

Oh then I misunderstood, Here the cursor does continue indeed, but recording (the transport's record button) is disabled very quickly, I thought that's the issue you're reporting.

**Comment by lowlou** [comment]
No ok sorry I wasn't looking the rec button, you are right, so yes it seems to be related. What happens here is less immediate, there is quarteer/half a second happening between the start and stop. But it's annonying, the recording process gets canceled, I have to start again, and when it happens once, it keeps happening unless I change the time selection, cursor etc. Very weird thing ! Might be Reaper-related indeed !

**Comment by lowlou** [comment]
Ok it seems that I found the problem. It comes from the fact that I was using the "wait n second" script in another cycle action before lauching  the recording cycle action. For some reason the "wait n second" script that I use in the recording action (so it's the second one used ok ?), it uses some infos of the first action, not infos of the second cycle action.  Basically it redoes the script of the first action (same time selection, cursor position). It's super weird to be honest. Stranger Things ++ !! But I've deleted the conflicting script in the first cycle action (it was not mandatory there), and the second cycle action seems stable now. I've launched 20 recordings in a row without problem :).

**Comment by nofishonfriday** [comment]
Nice you found the problem.
So closing this issue as 'solved'. :)
